### PR TITLE
Getting OAuth1 token

### PR DESCRIPTION
**Pull request by przywartya**

### CODE REVIEW

File: tap_twitter_ads/__init__.py [hunk 1]
```diff
@@ -20,10 +23,34 @@
...
+    if os.path.exists(credential_path):
+        with open(credential_path, "r") as file:
+            twitter_credentials = file.read()
+            if twitter_credentials:
+                credentials = json.loads(twitter_credentials)
+
+    if not credentials or credentials == "":
+        # Get new credentials
+
+    return credentials
```
Comment (paddie) [inline tap_twitter_ads/__init__.py]: While perfectly correct, using nested if-with-if like this makes it rather hard to reason about.

Consider instead using guard-clauses like below. It essentially amounts to having the happy path at the root of the function, and error checks etc. handled in ifs. This means you are essentially whittling the potential error cases away as you read further down. It is also easier for the eye to track on the vertical axis than the horizontal one.

Also, `None` is a perfectly reasonably "variable does not have a value yet" default. Although the below version does in fact not need it:

```suggestion
    if not os.path.exists(credential_path):
        return None
     
     with open(credential_path, "r") as file:
           return json.load(file)

    return None
```

File: tap_twitter_ads/__init__.py [hunk 2]
```diff
@@ -41,6 +41,18 @@ def get_credentials():
...
+        # Step 1: Obtain a request token which will identify you (the client) in the next step.
+        # At this stage you will only need your consumer key and secret.
+        oauth = OAuth1Session(
+            CONFIG["consumer_key"],
+            client_secret=CONFIG["consumer_secret"],
+            callback_uri="http://localhost:3000/oauth/validate",
+        )
+        fetch_response = oauth.fetch_request_token(CONFIG["request_token_url"])
+        resource_owner_key = fetch_response.get("oauth_token")
+        resource_owner_secret = fetch_response.get("oauth_token_secret")
```
Comment (paddie) [inline tap_twitter_ads/__init__.py]: this would then be moved to the left one level :)

File: tap_twitter_ads/__init__.py [hunk 3]
```diff
@@ -41,7 +41,7 @@ def get_credentials():
 
     if not credentials or credentials == "":
         # Get new credentials
-        # credentials_file = open(credential_path, "w+")
+        credentials_file = open(credential_path, "w+")
```
Comment (paddie) [inline tap_twitter_ads/__init__.py]: move this much closer to where it is needed

File: tap_twitter_ads/__init__.py [hunk 4]
```diff
@@ -68,6 +68,27 @@ def get_credentials():
...
+        )
+        oauth_tokens = oauth.fetch_access_token(CONFIG["access_token_url"])
+
+        new_credentials = {
+            "oauth_token": oauth_tokens.get("oauth_token"),
+            "oauth_token_secret": oauth_tokens.get("oauth_token_secret"),
+        }
+
+        credentials_file.write(json.dumps(new_credentials))
+        credentials_file.close()
```
Comment (paddie) [inline tap_twitter_ads/__init__.py]: by not doing the `credentials_file` further up, and instead doing it here, you can do this using `with` and making it a bit cleaner.

File: tap_twitter_ads/__init__.py [hunk 5]
```diff
@@ -30,7 +30,8 @@ def get_credentials():
     """ Get credentials """
 
     credential_path = os.path.join(
-        os.path.dirname(os.path.abspath(__file__)), CREDENTIALS_FILENAME
+        os.path.dirname(os.path.dirname(os.path.abspath(__file__))),
+        CREDENTIALS_FILENAME,
```
Comment (paddie) [inline tap_twitter_ads/__init__.py]: the only danger with this really, and we'd not have it once we remove the token stuff, is that we'd leave behind a file on the machine with some secret credentials. I get it is for debugging though, but we'd rather not leave files behind.

**Comment by przywartya** [comment]
> Some more style things to think about. In general, this function is _huge_. Try and break it into logical smaller parts like we did the with Bing stuff. I'd recommend the function to be a few lines, that call a function for each of the API calls.
> 
> Regardless, nice breakup of the commits 👍

Different people, different coding styles with breaking up code to smaller functions :p What's the point of having a small function if you use it only in one place :p I get it, readability etc. But there is no one perfect way to to this.

Food for thought: https://www.youtube.com/watch?v=Bks59AaHe1c

**Pull request merged**